### PR TITLE
DEV-2068: Allow advanced search filtering by TAS

### PR DESCRIPTION
**Pull request by kbard**
**Description:**

Provide for the ability to filter advanced search results by TAS components.

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [x] API documentation updated
3. [x] Necessary PR reviewers:
    - [x] Backend
    - [x] Frontend
4. [x] Matview impact assessment completed
5. [x] Frontend impact assessment completed
6. [x] Data validation completed
7. [x] Appropriate Operations ticket(s) created [DEV-3073](https://federal-spending-transparency.atlassian.net/browse/DEV-3073)
8. [x] Jira Ticket [DEV-2068](https://federal-spending-transparency.atlassian.net/browse/DEV-2068):
    - [x] Link to this Pull-Request

### CODE REVIEW

File: usaspending_api/accounts/helpers.py [hunk 1]
```diff
@@ -16,3 +31,33 @@ def start_and_end_dates_from_fyq(fiscal_year, fiscal_quarter):
...
+        return queryset.extra(
+            tables=[TASAwardMatview._meta.db_table],
+            where=[
+                '"{}".award_id = "{}".award_id'.format(TASAwardMatview._meta.db_table, model._meta.db_table),
+                where_sql,
+            ],
+            params=where_params
+        )
+
+    return queryset
```
Comment (BrianZito) [inline usaspending_api/accounts/helpers.py]: Why is this in the accounts folder and not awards or search?

Comment (emily-brents) [reply usaspending_api/accounts/helpers.py]: presumably because TAS = Treasury **Account** Symbol

Comment (kbard) [reply usaspending_api/accounts/helpers.py]: I don't know.  It's very confusing.  There's stuff scattered everywhere for this set of endpoints/models.  I just put it with the constants.  Once again, I don't know how much I care.  If you have a preference, let it be known... awards or search or somewhere else entirely.

Comment (BrianZito) [reply usaspending_api/accounts/helpers.py]: I think it should be in search for cleaner encapsulation. I can do it myself if that would be easier for you (and won't make a bunch of code conflicts).

File: usaspending_api/api_contracts/contracts/search/AdvancedSearch.md [hunk 2]
```diff
@@ -319,6 +319,7 @@ List of table columns
 + `contract_pricing_type_codes`: `J` (optional, array[string])
 + `set_aside_type_codes`: `NONE` (optional, array[string])
 + `extent_competed_type_codes`: `A` (optional, array[string])
++ `tas_codes` (optional, array[TASCodeObject], fixed-type)
```
Comment (tony-sappe) [inline usaspending_api/api_contracts/contracts/search/AdvancedSearch.md]: These contracts need to include this new filter key and object

`usaspending-api/usaspending_api/api_contracts/contracts/search/SpendingByAwardCount.md` 
`usaspending-api/usaspending_api/api_contracts/contracts/search/SpendingByGeography.md`

File: usaspending_api/api_contracts/contracts/search/AdvancedSearch.md [hunk 3]
```diff
@@ -348,6 +347,22 @@ List of table columns
...
+    Allocation Transfer Agency Identifier - three characters
++ `aid` (required, string)
+    Agency Identifier - three characters
++ `bpoa` (optional, string, nullable)
+    Beginning Period of Availability - four digits
++ `epoa` (optional, string, nullable)
+    Ending Period of Availability - four digits
++ `a` (optional, string, nullable)
+    Availability Type Code - X or null
++ `main` (required, string)
```
Comment (tony-sappe) [inline usaspending_api/api_contracts/contracts/search/AdvancedSearch.md]: Question: is `main` used for Federal Account filtering?

Comment (kbard) [reply usaspending_api/api_contracts/contracts/search/AdvancedSearch.md]: Yes.  Federal Account is `aid` and `main` (and technically `sub`, but my understanding is that we're removing that from the front end).

File: usaspending_api/common/validator/award_filter.py [hunk 4]
```diff
@@ -86,6 +86,20 @@
...
+        "name": "tas_codes",
+        "type": "array",
+        "array_type": "object",
+        "object_keys": {
+            "ata": {"type": "text", "text_type": "search", "optional": True, "allow_nulls": True},
+            "aid": {"type": "text", "text_type": "search", "optional": False, "allow_nulls": False},
+            "bpoa": {"type": "text", "text_type": "search", "optional": True, "allow_nulls": True},
+            "epoa": {"type": "text", "text_type": "search", "optional": True, "allow_nulls": True},
+            "a": {"type": "text", "text_type": "search", "optional": True, "allow_nulls": True},
+            "main": {"type": "text", "text_type": "search", "optional": False, "allow_nulls": False},
```
Comment (tony-sappe) [inline usaspending_api/common/validator/award_filter.py]: shouldn't this be required? (optional: False)

File: usaspending_api/database_scripts/matview_generator/tas_award_matview.json [hunk 5]
```diff
@@ -0,0 +1,64 @@
+{
+  "final_name": "tas_award_matview",
+  "refresh": true,
+  "matview_sql": [
+    "select distinct",
+    "    md5(concat_ws(",
+    "        '|',",
```
Comment (tony-sappe) [inline usaspending_api/database_scripts/matview_generator/tas_award_matview.json]: What benefit does the pipe character  provide when creating the MD5 UUID?

Comment (kbard) [reply usaspending_api/database_scripts/matview_generator/tas_award_matview.json]: You mean in comparison to some other character or no character at all?  It's no more or less valuable than a space in this particular case.  I'd prefer to use some kind of separator character to minimize the chance of collisions.

Comment (kbard) [reply usaspending_api/database_scripts/matview_generator/tas_award_matview.json]: By way of totally contrived example, the following two values would hash to the same thing without some sort of delimiter (preferably one that "cannot" exist in the actual data):

| aid | a | main | sub | transaction_id |
|-----|---|-------|-----|-----------------|
| 011  |  X | 1234  |       |  12345678         |
| 011  |  X | 1234  | 123  |  45678              |

Comment (tony-sappe) [reply usaspending_api/database_scripts/matview_generator/tas_award_matview.json]: Thanks for the example, it makes complete sense to me.

**Pull request merged**